env_process: Handle the kvm_probe module together

PaulYuuu opened the pull request:
1. arch.get_kvm_module_list: Update kvm module list for powerpc and s390
2. Add support to handle the kvm_probe module.
3. Improve "reload_module" and "restore" methods which provided by
utils_kernel_module, now they will unload the module and those used by
this one.

ID: 1836116, 1836123
Signed-off-by: Yihuang Yu <yihyu@redhat.com>

PaulYuuu commented:
@luckyh @smitterl Can you help to review? Thanks a lot!

YongxueHong commented on virttest/utils_kernel_module.py:
```diff
@@ -134,8 +139,13 @@ def restore(self):
+                process.run(unload_cmd + ' --dry-run')
+                cmds.append(unload_cmd)
+            except process.CmdError:
+                cmds.append('modprobe -r %s' % self._module_name)
```
Just suggestion:Might we pick up them into a private function to avoid redundantly.

PaulYuuu replied:
Thanks, introduced a new method.

luckyh commented on virttest/env_process.py:
```diff
@@ -1174,7 +1175,10 @@ def preprocess(test, params, env):
     global KVM_MODULE_HANDLER
-    KVM_MODULE_HANDLER = utils_kernel_module.reload("kvm", force_load, kvm_module_params)
+    # Only unload/reload the kvm_probe module is required
+    kvm_module = arch.get_kvm_module_list()[-1].replace('-', '_')
```
Instead of replacing with the arch specified module directly, it's better to introduce another routine to handle the modification of arch modules.

PaulYuuu commented:
> Hi @PaulYuuu . Thank you for this enhancement and also for adding unit tests :)
> There are some things that worry me. When we added the kernel module handler we had two goals that were important for us:
> 
> 1. Delegate all logic regarding configuration, object creation etc. to the utils_kernel_module.py
> 2. Have the least amount of function calls possible
> 
@smitterl thanks your review. Yes, this is also my worry that:
```python
if module_used_by:
    for module in module_used_by.group(1).split(','):
        self._module_used_by.append(KernelModuleHandler(module))
```
> In this PR there's a mixed approach with two module handlers in env_process and one of them (for kvm) having an additional third handler internally to handle kvm-hv.
> 
> Maybe we can find a solution so that 1. and 2. still hold.
> 
> I observed on x86_64 with (kvm, kvm_intel) two things that might be useful:
> 
> 1. It doesn't matter if 'nested=1' is defined for modprobe of kvm or kvm_intel.

Not completely accurate. In RHEL7, it does not work and the exit code is `1`. A difference in RHEL8 is that the exit code returns `0`. (Will ignore invalid parameters)

RHEL7:
```bash
$ modprobe -r kvm_intel
$ modprobe -r kvm
$ modprobe kvm nested=1
modprobe: ERROR: could not insert 'kvm': Unknown symbol in module, or unknown parameter (see dmesg)
$ echo $?
1
$ dmesg | grep nested
[13504.531214] kvm: Unknown parameter `nested'
[13532.795956] kvm: Unknown parameter `nested'
$ 
```

RHEL8:
```bash
$ modprobe -r kvm_intel
$ modprobe -r kvm
$ modprobe kvm nested=1
$ echo $?
0
$ dmesg | grep nested
[259197.076366] kvm: unknown parameter 'nested' ignored
[260025.315738] kvm: unknown parameter 'nested' ignored
$ 
```

> 2. Defining it for kvm_intel automatically loads kvm, too; that is 'modprobe -r kvm_intel; modprobe -r kvm; modprobe kvm_intel nested=1; lsmode|grep kvm ==> kvm_intel has been loaded with nested=1 and kvm has been loaded.
> 

Yes, when `kvm_intel` has been loaded, it will automatically loads `kvm` module, but if we want to load these two module with some parameters, we must keep them independent, otherwise, `modprobe -r kvm_intel` will unload `kvm` together and restore those parameters we want to set.
> The following are two proposals based on your approach that I think might work:
> 
> I.
> With [autotest/tp-qemu#2230](https://github.com/autotest/tp-qemu/pull/2230) in mind, using 1. and based on your approach it might be enough to define 'nested=1' for kvm and add only the logic necessary to kernel_module_handler to
> a. unload modules that use it before reloading (IF it should be reloaded, same logic as before)
> b. load kvm with the parameters as before followed by loading the modules that were onloaded in a.
> 
> The flow would be something like:
> 
> ```
> env_process.py:
> 
>     KVM_MODULE_HANDLER = utils_kernel_module.reload("kvm", force, params):
> 
> utils_kernel_module.py:
> 
> ==> kvm_module_handler = utils_kernel_module.__init__("kvm") # identifies 'kvm-hv' as using 'kvm'
> ===> kvm_module_handler.reload_module(force, params)
> ====> if not force:
>         ...
>         self._unload_using_modules() # unloads 'kvm-hv'
>         self.reload(force, params) # loads 'kvm' with 'nested=1'
>         self._load_using_modules() # loads 'kvm-hv'
> ```
> 
> and env_process wouldn't need to be changed. The `restore` method should then unload/load using modules accordingly like shown above.
> 
> II.
> There's also another approach I can think of that allows for explicit module configurations and order; `utils_module_kernel.reload` would have a different signature and several module configs are defined in a single params string and the actual KernelModuleHandler wouldn't have to track the modules / reload the modules using it:

If so, I think `def __init__(self, module_name):` should be an iterator, so that this object can handle all related modules.
> 
> ```
> env_process.py:
>     KVM_MODULE_HANDLERS = utils_kernel_module.reload("kvm-hv:nested=1;kvm:", force)
> 
> utils_kernel_module.py/reload:
> 
> ==> if params != "" or force:
> ===> module_params = params.split(";")
>      handlers = []
>      for mp in module_params:
>        mname, mparams = mp.split(":")
>        handler = KernelModuleHandler(mname)
>        handler.reload_module(force, mparams)
>        handlers.append(handler)
> 
> env_process.py:
> 
> if KVM_MODULE_HANDLERS:
>    utils_kernel_module.restore(KVM_MODULE_HANDLERS) # restores in order of list, that is first 'kvm-hv' then 'kvm'
> ```
> 
> I believe II. is probably more useful as it doesn't restrict too much on this very use case of defining nested=1 for kvm-hv. What do you think?

PaulYuuu commented on virttest/arch.py:
```diff
@@ -89,4 +89,6 @@ def get_kvm_module_list():
         return ["kvm", "kvm-%s" % host_cpu_type]
     elif ARCH in ('ppc64', 'ppc64le'):
+        return ["kvm", "kvm-hv"]
+    elif ARCH in ('s390', 's390x'):
```
Yes, this may a problem on POWER8, if the `kvm_hv` module is not loaded, `virttest/bootstrap.py` will report a warning.
BTW, `kvm_pr` has not parameters to set.
```bash
$ ls /sys/module/kvm_pr/
coresize  initsize   notes   rhelversion  srcversion  uevent
holders   initstate  refcnt  sections     taint
```

smitterl replied:
> Yes, this may a problem on POWER8, if the `kvm_hv` module is not loaded, `virttest/bootstrap.py` will report a warning.
> BTW, `kvm_pr` has not parameters to set.
> 
> ```shell
> $ ls /sys/module/kvm_pr/
> coresize  initsize   notes   rhelversion  srcversion  uevent
> holders   initstate  refcnt  sections     taint
> ```

If you use the design II. taking a string like `kvm-hv:nested=1;kvm:` as input for `.reload`, could usage of `get_kvm_module_list` replaced by `kvm-pr:;kvm:param1=val1`?

smitterl commented:
> > Hi @PaulYuuu . Thank you for this enhancement and also for adding unit tests :)
> > There are some things that worry me. When we added the kernel module handler we had two goals that were important for us:
> > 
> > 1. Delegate all logic regarding configuration, object creation etc. to the utils_kernel_module.py
> > 2. Have the least amount of function calls possible
> 
> @smitterl thanks your review. Yes, this is also my worry that:
> 
> ```python
> if module_used_by:
>     for module in module_used_by.group(1).split(','):
>         self._module_used_by.append(KernelModuleHandler(module))
> ```
> 
> > In this PR there's a mixed approach with two module handlers in env_process and one of them (for kvm) having an additional third handler internally to handle kvm-hv.
> > Maybe we can find a solution so that 1. and 2. still hold.
> > I observed on x86_64 with (kvm, kvm_intel) two things that might be useful:
> > 
> > 1. It doesn't matter if 'nested=1' is defined for modprobe of kvm or kvm_intel.
> 
> Not completely accurate. In RHEL7, it does not work and the exit code is `1`. A difference in RHEL8 is that the exit code returns `0`. (Will ignore invalid parameters)

Got it. Thank you :smiley: 
I conclude that I. doesn't make sense and we should opt for something like II.

> > II.
> > There's also another approach I can think of that allows for explicit module configurations and order; `utils_module_kernel.reload` would have a different signature and several module configs are defined in a single params string and the actual KernelModuleHandler wouldn't have to track the modules / reload the modules using it:
> 
> If so, I think `def __init__(self, module_name):` should be an iterator, so that this object can handle all related modules.

I'm not sure I understand what you mean. For II., I imagined each KernelModuleHandler instance to handle exactly one module and no change to the class would actually be needed while the module level functions `restore` and `reload` would take care of handling the ordered execution; i.e. the order of the module names in the params string `"kvm-hv:nested=1;kvm:"` is important:

>> ```python
>> env_process.py:
>>    KVM_MODULE_HANDLERS = utils_kernel_module.reload("kvm-hv:nested=1;kvm:", force)
>>
>> utils_kernel_module.py/reload:
>>
>> ==> if params != "" or force:
>> ===> module_params = params.split(";")
>>      handlers = []
>>      for mp in module_params: # this for loop first calls reload_module for 'kvm-hv', then for 'kvm'
>>        mname, mparams = mp.split(":")
>>        handler = KernelModuleHandler(mname)
>>        handler.reload_module(force, mparams)
>>        handlers.append(handler) # this establishes the order having 'kvm-hv' as first element
>>
>>      return handlers
>>
>> env_process.py:
>> 
>> if KVM_MODULE_HANDLERS:
>>    utils_kernel_module.restore(KVM_MODULE_HANDLERS) # restores in order of list, that is first 'kvm-hv' then 'kvm'
>> ```

``` python
utils_kernel_module.py/restore(module_handlers):
...
  # will first call restore on 'kvm-hv' then on 'kvm' because this is the order
  # in which the handlers have been added to the module_handlers list in the 
  # utils_kernel_module.reload function (s. above)
  for handler in module_handlers:
      handler.restore()
```

PaulYuuu commented on virttest/utils_kernel_module.py:
```diff
@@ -17,13 +17,46 @@
 
 
-def reload(module_name, force, params):
+class KernelModuleError(Exception):
```
Not completely agree, this is at the environmental preparation stage, the test case has not been executed yet.

smitterl replied:
@PaulYuuu It seemed to me that this was the best choice of avocado's exceptions https://github.com/avocado-framework/avocado/blob/master/avocado/core/exceptions.py.

PaulYuuu commented on virttest/env_process.py:
```diff
@@ -1171,10 +1173,19 @@ def preprocess(test, params, env):
+        if len(kvm_modules) == 2:
+            KVM_PROBE_MODULE_HANDLER = utils_kernel_module.reload(
+                kvm_modules[1], kvm_probe_force_load, kvm_probe_module_params)
+        KVM_MODULE_HANDLER = utils_kernel_module.reload(
```
It works after applying this patch.
```
2020-07-02 05:42:24,402 utils_kernel_mod L0217 DEBUG| Backed up kvm-hv module state (was_loaded, params)=(True, dynamic_mt_modes=6 target_smt_mode=0 one_vm_per_core=N indep_threads_mode=Y nested=Y kvm_irq_bypass=1 h_ipi_redirect=1)
2020-07-02 05:42:24,408 utils_kernel_mod L0110 DEBUG| Unloading module: modprobe -r kvm-hv
2020-07-02 05:42:24,701 utils_kernel_mod L0161 DEBUG| Reloading module: modprobe kvm-hv nested=1
2020-07-02 05:42:24,741 utils_kernel_mod L0217 DEBUG| Backed up kvm module state (was_loaded, params)=(True, halt_poll_ns_shrink=0 halt_poll_ns=10000 halt_poll_ns_grow_start=10000 halt_poll_ns_grow=2)
2020-07-02 05:42:24,747 utils_kernel_mod L0217 DEBUG| Backed up kvm_hv module state (was_loaded, params)=(True, dynamic_mt_modes=6 target_smt_mode=0 one_vm_per_core=N indep_threads_mode=Y nested=Y kvm_irq_bypass=1 h_ipi_redirect=1)
2020-07-02 05:42:24,752 utils_kernel_mod L0110 DEBUG| Unloading module: modprobe -r kvm_hv
2020-07-02 05:42:25,081 utils_kernel_mod L0110 DEBUG| Unloading module: modprobe -r kvm
2020-07-02 05:42:25,084 utils_kernel_mod L0161 DEBUG| Reloading module: modprobe kvm
2020-07-02 05:42:25,112 utils_kernel_mod L0195 DEBUG| Restoring module state: modprobe kvm_hv dynamic_mt_modes=6 target_smt_mode=0 one_vm_per_core=N indep_threads_mode=Y nested=Y kvm_irq_bypass=1 h_ipi_redirect=1
```

PaulYuuu commented on virttest/utils_kernel_module.py:
```diff
@@ -48,12 +81,38 @@ def __init__(self, module_name):
+                                         self._module_name.replace('-', '_'))
+        self._module_params_path = os.path.join(self._module_path, 'parameters')
         self._was_loaded = None
-        self._loaded_config = None
```
I removed it because we have backup the config, check it dynamically is more reasonable, if they do not match, restore it. Otherwise, when reloading `kvm`, the kvm probe module failed to reload.
But since we are talking about the current design, I will take care of it.

PaulYuuu commented on virttest/unittests/test_utils_kernel_module.py:
```diff
@@ -132,26 +162,24 @@ class TestRestore(unittest.TestCase):
         self.assertTrue(getstatusoutput_ok.called)
-        cmd = getstatusoutput_ok.call_args[0][0]
-        self.assertTrue(params in cmd)
+        self.assertTrue(orig_params == cur_params)
```
Good catch, I will modify it.

PaulYuuu commented on virttest/utils_kernel_module.py:
```diff
@@ -48,12 +81,38 @@ def __init__(self, module_name):
         self._backup_config()
 
-    def reload_module(self, force, params):
+        search_pattern = r'^%s\s+\d+\s+\d\s+(\S+)$' % self._module_name
```
It's okay for me.

PaulYuuu commented on virttest/env_process.py:
```diff
@@ -1171,10 +1173,19 @@ def preprocess(test, params, env):
+    global KVM_PROBE_MODULE_HANDLER
     global KVM_MODULE_HANDLER
-    KVM_MODULE_HANDLER = utils_kernel_module.reload("kvm", force_load, kvm_module_params)
+    if arch.ARCH != "aarch64":
```
ARM has not any kvm module, so return an empty list is ok, and if do that, I think better to define the `KVM_MODULE_HANDLER` as a list to handle all modules returned by `get_kvm_module_list`.

PaulYuuu commented on virttest/utils_kernel_module.py:
```diff
@@ -48,12 +81,38 @@ def __init__(self, module_name):
+
+        If there are some modules using this module, will unload them at first
+        and then unload this one.
+        """
```
Thank you for providing a clearer description.

PaulYuuu commented:
@smitterl Sorry for the delay.
> I imagined each KernelModuleHandler instance to handle exactly one module and no change to the class would actually be needed while the module level functions restore and reload would take care of handling the ordered execution.

I have considered this method, it's an easy way to handle all modules, and only need to add a method `def unload()` in `KernelModuleHandler`.
```
KVM_MODULE_HANDLER = KernelModuleHandler("kvm")
KVM_PROBE_MODULE_HANDLER = KernelModuleHandler("kvm_hv")

KVM_PROBE_MODULE_HANDLER.unload()
KVM_MODULE_HANDLER.reload()
KVM_PROBE_MODULE_HANDLER.reload()
```

However, users need to pay attention to the dependencies between modules themselves, but not so now. @luckyh what do you think?

PaulYuuu commented:
@smitterl Any comments?
For the current design, it's an easy way to handle the module and its dependence.
The second one is to split this class to handle a single module handler, if fail to unload/reload it, raise the exception directly.
Code should like:
```python
# pre_process
KVM_MODULE_HANDLER = KernelModuleHandler("kvm")
KVM_PROBE_MODULE_HANDLER = KernelModuleHandler("kvm_hv")
KVM_PROBE_MODULE_HANDLER.unload()
KVM_MODULE_HANDLER.reload()
KVM_PROBE_MODULE_HANDLER.reload()

# post_process
KVM_PROBE_MODULE_HANDLER.unload()
KVM_MODULE_HANDLER.restore()
KVM_PROBE_MODULE_HANDLER.restore()
```

And if you have a better idea, we can pick that one, thanks.

PaulYuuu commented:
Scenario (1/7): `kvm_module_force_load = yes`
```
04:56:23 DEBUG| Backed up kvm module state (was_loaded, params)=(True, halt_poll_ns_shrink=0 halt_poll_ns=10000 halt_poll_ns_grow_start=10000 halt_poll_ns_grow=2)
04:56:23 DEBUG| Backed up kvm_hv module state (was_loaded, params)=(True, dynamic_mt_modes=6 target_smt_mode=0 one_vm_per_core=N indep_threads_mode=Y nested=Y kvm_irq_bypass=1 h_ipi_redirect=1)
04:56:23 DEBUG| Unloading module: rmmod kvm_hv
04:56:23 DEBUG| Unloading module: rmmod kvm
04:56:23 DEBUG| Reloading module: modprobe kvm
04:56:24 DEBUG| Restoring module state: modprobe kvm_hv dynamic_mt_modes=6 target_smt_mode=0 one_vm_per_core=N indep_threads_mode=Y nested=Y kvm_irq_bypass=1 h_ipi_redirect=1
```

Scenario (2/7): `kvm_module_parameters = halt_poll_ns_shrink=1`
```
04:41:09 DEBUG| Backed up kvm module state (was_loaded, params)=(True, halt_poll_ns_shrink=0 halt_poll_ns=10000 halt_poll_ns_grow_start=10000 halt_poll_ns_grow=2)
04:41:09 DEBUG| Backed up kvm_hv module state (was_loaded, params)=(True, dynamic_mt_modes=6 target_smt_mode=0 one_vm_per_core=N indep_threads_mode=Y nested=Y kvm_irq_bypass=1 h_ipi_redirect=1)
04:41:09 DEBUG| Unloading module: rmmod kvm_hv
04:41:09 DEBUG| Unloading module: rmmod kvm
04:41:10 DEBUG| Reloading module: modprobe kvm halt_poll_ns_shrink=1
04:41:10 DEBUG| Restoring module state: modprobe kvm_hv dynamic_mt_modes=6 target_smt_mode=0 one_vm_per_core=N indep_threads_mode=Y nested=Y kvm_irq_bypass=1 h_ipi_redirect=1

04:53:19 DEBUG| Backed up kvm_hv module state (was_loaded, params)=(True, dynamic_mt_modes=6 target_smt_mode=0 one_vm_per_core=N indep_threads_mode=Y nested=Y kvm_irq_bypass=1 h_ipi_redirect=1)
04:53:19 DEBUG| Unloading module: rmmod kvm_hv
04:53:19 DEBUG| Unloading module: rmmod kvm
04:53:19 DEBUG| Restoring module state: modprobe kvm halt_poll_ns_shrink=0 halt_poll_ns=10000 halt_poll_ns_grow_start=10000 halt_poll_ns_grow=2
04:53:19 DEBUG| Restoring module state: modprobe kvm_hv dynamic_mt_modes=6 target_smt_mode=0 one_vm_per_core=N indep_threads_mode=Y nested=Y kvm_irq_bypass=1 h_ipi_redirect=1
```

Scenario (3/7): `kvm_probe_module_force_load = yes`
```
05:24:19 DEBUG| Backed up kvm-hv module state (was_loaded, params)=(True, dynamic_mt_modes=6 target_smt_mode=0 one_vm_per_core=N indep_threads_mode=Y nested=Y kvm_irq_bypass=1 h_ipi_redirect=1)
05:24:19 DEBUG| Unloading module: rmmod kvm-hv
05:24:19 DEBUG| Reloading module: modprobe kvm-hv
```

Scenario (4/7): `kvm_probe_module_parameters = nested=1`
```
05:29:16 DEBUG| Backed up kvm-hv module state (was_loaded, params)=(True, dynamic_mt_modes=6 target_smt_mode=0 one_vm_per_core=N indep_threads_mode=Y nested=N kvm_irq_bypass=1 h_ipi_redirect=1)
05:29:16 DEBUG| Unloading module: rmmod kvm-hv
05:29:17 DEBUG| Reloading module: modprobe kvm-hv nested=1

05:31:00 DEBUG| Unloading module: rmmod kvm-hv
05:31:00 DEBUG| Restoring module state: modprobe kvm-hv dynamic_mt_modes=6 target_smt_mode=0 one_vm_per_core=N indep_threads_mode=Y nested=N kvm_irq_bypass=1 h_ipi_redirect=1
```

Scenario (5/7): `kvm_module_force_load = yes` `kvm_probe_module_parameters = nested=1`
```
05:32:52 DEBUG| Backed up kvm-hv module state (was_loaded, params)=(True, dynamic_mt_modes=6 target_smt_mode=0 one_vm_per_core=N indep_threads_mode=Y nested=N kvm_irq_bypass=1 h_ipi_redirect=1)
05:32:52 DEBUG| Unloading module: rmmod kvm-hv
05:32:52 DEBUG| Reloading module: modprobe kvm-hv nested=1
05:32:52 DEBUG| Backed up kvm module state (was_loaded, params)=(True, halt_poll_ns_shrink=0 halt_poll_ns=10000 halt_poll_ns_grow_start=10000 halt_poll_ns_grow=2)
05:32:52 DEBUG| Backed up kvm_hv module state (was_loaded, params)=(True, dynamic_mt_modes=6 target_smt_mode=0 one_vm_per_core=N indep_threads_mode=Y nested=Y kvm_irq_bypass=1 h_ipi_redirect=1)
05:32:52 DEBUG| Unloading module: rmmod kvm_hv
05:32:52 DEBUG| Unloading module: rmmod kvm
05:32:52 DEBUG| Reloading module: modprobe kvm
05:32:53 DEBUG| Restoring module state: modprobe kvm_hv dynamic_mt_modes=6 target_smt_mode=0 one_vm_per_core=N indep_threads_mode=Y nested=Y kvm_irq_bypass=1 h_ipi_redirect=1

05:34:37 DEBUG| Unloading module: rmmod kvm-hv
05:34:37 DEBUG| Restoring module state: modprobe kvm-hv dynamic_mt_modes=6 target_smt_mode=0 one_vm_per_core=N indep_threads_mode=Y nested=N kvm_irq_bypass=1 h_ipi_redirect=1
```

Scenario (6/7): `kvm_probe_module_force_load = yes` `kvm_module_parameters = halt_poll_ns_shrink=1`
```
05:44:39 DEBUG| Backed up kvm-hv module state (was_loaded, params)=(True, dynamic_mt_modes=6 target_smt_mode=0 one_vm_per_core=N indep_threads_mode=Y nested=N kvm_irq_bypass=1 h_ipi_redirect=1)
05:44:39 DEBUG| Unloading module: rmmod kvm-hv
05:44:39 DEBUG| Reloading module: modprobe kvm-hv
05:44:39 DEBUG| Backed up kvm module state (was_loaded, params)=(True, halt_poll_ns_shrink=0 halt_poll_ns=10000 halt_poll_ns_grow_start=10000 halt_poll_ns_grow=2)
05:44:39 DEBUG| Backed up kvm_hv module state (was_loaded, params)=(True, dynamic_mt_modes=6 target_smt_mode=0 one_vm_per_core=N indep_threads_mode=Y nested=Y kvm_irq_bypass=1 h_ipi_redirect=1)
05:44:39 DEBUG| Unloading module: rmmod kvm_hv
05:44:40 DEBUG| Unloading module: rmmod kvm
05:44:40 DEBUG| Reloading module: modprobe kvm halt_poll_ns_shrink=1
05:44:40 DEBUG| Restoring module state: modprobe kvm_hv dynamic_mt_modes=6 target_smt_mode=0 one_vm_per_core=N indep_threads_mode=Y nested=Y kvm_irq_bypass=1 h_ipi_redirect=1

05:46:23 DEBUG| Unloading module: rmmod kvm-hv
05:46:23 DEBUG| Restoring module state: modprobe kvm-hv dynamic_mt_modes=6 target_smt_mode=0 one_vm_per_core=N indep_threads_mode=Y nested=N kvm_irq_bypass=1 h_ipi_redirect=1
05:46:23 DEBUG| Backed up kvm_hv module state (was_loaded, params)=(True, dynamic_mt_modes=6 target_smt_mode=0 one_vm_per_core=N indep_threads_mode=Y nested=N kvm_irq_bypass=1 h_ipi_redirect=1)
05:46:23 DEBUG| Unloading module: rmmod kvm_hv
05:46:23 DEBUG| Unloading module: rmmod kvm
05:46:23 DEBUG| Restoring module state: modprobe kvm halt_poll_ns_shrink=0 halt_poll_ns=10000 halt_poll_ns_grow_start=10000 halt_poll_ns_grow=2
05:46:23 DEBUG| Restoring module state: modprobe kvm_hv dynamic_mt_modes=6 target_smt_mode=0 one_vm_per_core=N indep_threads_mode=Y nested=N kvm_irq_bypass=1 h_ipi_redirect=1
```

Scenario (7/7): `kvm_module_parameters = halt_poll_ns_shrink=1` `kvm_probe_module_parameters = nested=1`
```
05:50:47 DEBUG| Backed up kvm-hv module state (was_loaded, params)=(True, dynamic_mt_modes=6 target_smt_mode=0 one_vm_per_core=N indep_threads_mode=Y nested=Y kvm_irq_bypass=1 h_ipi_redirect=1)
05:50:47 DEBUG| Unloading module: rmmod kvm-hv
05:50:47 DEBUG| Reloading module: modprobe kvm-hv nested=1
05:50:47 DEBUG| Backed up kvm module state (was_loaded, params)=(True, halt_poll_ns_shrink=0 halt_poll_ns=10000 halt_poll_ns_grow_start=10000 halt_poll_ns_grow=2)
05:50:47 DEBUG| Backed up kvm_hv module state (was_loaded, params)=(True, dynamic_mt_modes=6 target_smt_mode=0 one_vm_per_core=N indep_threads_mode=Y nested=Y kvm_irq_bypass=1 h_ipi_redirect=1)
05:50:47 DEBUG| Unloading module: rmmod kvm_hv
05:50:48 DEBUG| Unloading module: rmmod kvm
05:50:48 WARNI| IPv6 address sniffing is not supported yet by using TShark, please fallback to use other sniffers by uninstalling TShark when testing with IPv6
05:50:48 DEBUG| Reloading module: modprobe kvm halt_poll_ns_shrink=1
05:50:48 DEBUG| Restoring module state: modprobe kvm_hv dynamic_mt_modes=6 target_smt_mode=0 one_vm_per_core=N indep_threads_mode=Y nested=Y kvm_irq_bypass=1 h_ipi_redirect=1

05:52:33 DEBUG| Backed up kvm_hv module state (was_loaded, params)=(True, dynamic_mt_modes=6 target_smt_mode=0 one_vm_per_core=N indep_threads_mode=Y nested=Y kvm_irq_bypass=1 h_ipi_redirect=1)
05:52:33 DEBUG| Unloading module: rmmod kvm_hv
05:52:33 DEBUG| Unloading module: rmmod kvm
05:52:33 DEBUG| Restoring module state: modprobe kvm halt_poll_ns_shrink=0 halt_poll_ns=10000 halt_poll_ns_grow_start=10000 halt_poll_ns_grow=2
05:52:33 DEBUG| Restoring module state: modprobe kvm_hv dynamic_mt_modes=6 target_smt_mode=0 one_vm_per_core=N indep_threads_mode=Y nested=Y kvm_irq_bypass=1 h_ipi_redirect=1
```

PaulYuuu commented:
@YongxueHong Could you also help to review? Thanks a lot.

YongxueHong commented:
Hi @PaulYuuu 
Look good to me, thanks.

luckyh commented on virttest/utils_kernel_module.py:
```diff
@@ -132,22 +177,21 @@ def restore(self):
-            cmds = []
-            # TODO: Handle cases were module cannot be removed
-            cmds.append('modprobe -r --remove-dependencies %s' % self._module_name)
+        if self._get_serialized_config() != self._config_backup:
```
Looks `if self.current_config != self._config_backup:` is better.

PaulYuuu replied:
Yes, I added its property, corrected it.

luckyh commented on virttest/utils_kernel_module.py:
```diff
@@ -132,22 +177,21 @@ def restore(self):
 
-        if self._loaded_config is not None:
-            cmds = []
-            # TODO: Handle cases were module cannot be removed
```
Better to keep this comment since we still not solve the issue.

PaulYuuu replied:
Restored.

luckyh commented on virttest/utils_kernel_module.py:
```diff
@@ -97,20 +145,17 @@ def reload_module(self, force, params):
 
-        cmds = []
-        if self._was_loaded:
-            # TODO: Handle cases were module cannot be removed
```
+1

luckyh commented on virttest/arch.py:
```diff
@@ -89,4 +89,8 @@ def get_kvm_module_list():
         host_cpu_type = cpu.get_cpu_vendor_name()
         return ["kvm", "kvm-%s" % host_cpu_type]
     elif ARCH in ('ppc64', 'ppc64le'):
+        return ["kvm", "kvm-hv"]
```
Similar as x86, it's possible to detect the probe module dynamically, so do you want add it together? if no, let's leave a TODO comment for that, thanks!

PaulYuuu replied:
Since vt only support POWER8 and POWER9, and they all support kvm_hv, so I only add a comment here.

luckyh commented:
Let's merge this since the kvm probe modules can be configured well now. However, as @smitterl pointed out, there are still redundant actions for loading/unloading modules, so please follow up with other enhancements if necessary, thanks.
